feat: Implement Spark unhex

tshauck opened the pull request:
## Which issue does this PR close?

Closes #341

## Rationale for this change

`unhex` is currently unsupported by comet. This my first PR into this repo, so certainly open to any feedback to make it more inline w/ expectations.

## What changes are included in this PR?

Add `unhex` as well as make some minor refactors.

## How are these changes tested?

Added simple tests to the rust and spark sql side of the code.

tshauck commented on core/src/execution/datafusion/planner.rs:
```diff
@@ -1301,24 +1301,26 @@ impl PhysicalPlanner {
-                let fun = BuiltinScalarFunction::from_str(fun_name);
-                if fun.is_err() {
+
+                if let Ok(fun) = BuiltinScalarFunction::from_str(fun_name) {
```
Unrelated, but more idiomatic IMO

tshauck commented on core/src/execution/datafusion/expressions/scalar_funcs.rs:
```diff
@@ -123,11 +126,10 @@ pub fn create_comet_physical_fun(
-            let fun = BuiltinScalarFunction::from_str(fun_name);
-            if fun.is_err() {
-                Ok(ScalarFunctionDefinition::UDF(registry.udf(fun_name)?))
+            if let Ok(fun) = BuiltinScalarFunction::from_str(fun_name) {
```
Unrelated, but more idiomatic IMO

tshauck commented on spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:
```diff
@@ -1396,6 +1396,17 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde {
           val optExpr = scalarExprToProto("atan2", leftExpr, rightExpr)
           optExprWithInfo(optExpr, expr, left, right)
 
+        case e @ Unhex(child, failOnError) =>
```
https://github.com/apache/spark/blob/59d5946cfd377e9203ccf572deb34f87fab7510c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala#L1152
vs 
https://github.com/apache/spark/blob/45ba9224602eb18fe45e339cbb8cf2e8a4924f0b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala#L1125

andygrove replied:
We need to use a shim approach. We currently have one folder with source for spark-3.x. We should add additional folders as needed e.g. spark-3.3.x / spark-3.4.x with any code that is specific to those versions.

I am happy to help with this.

kazuyukitanimura replied:
I think we can handle this using existing shim, something like
```
case e @ Unhex =>
e.child...
getFailOnError(e) ...
```

kazuyukitanimura replied:
I was wrong, actually `getFailOnError` needs to be modified for this approach as `unhex` is `unaryExpression`

tshauck replied:
@andygrove Would you mind having a look at the updated PR and let me know what you think of the shim implemntation? The thought was to setup a major and minor shim to keep the existing shim for all of spark 3 then apply the more granular difference in the minor. It gets a little weird in that I'm applying 3.3's shim to 3.2.

@kazuyukitanimura Thanks for looking!

andygrove replied:
Thanks @tshauck I will review this tomorrow.

kazuyukitanimura replied:
I think we can have duplicated shim for 3.3 and 3.2

kazuyukitanimura commented on spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:
```diff
@@ -45,12 +45,13 @@ import org.apache.comet.CometSparkSessionExtensions.{isCometOperatorEnabled, isC
  * An utility object for query plan and expression serialization.
  */
-object QueryPlanSerde extends Logging with ShimQueryPlanSerde {
+object QueryPlanSerde extends Logging with ShimQueryPlanSerde with ShimCometUnhexExpr {
```
I would say `ShimCometUnhexExpr` should be more generic name, not unhex specific. What about `ShimCometExpr ` (or if you have other idea, please feel free to propose)?
Otherwise, we will have to keep adding trait class per function.

Eventually we should merge  `ShimQueryPlanSerde` and `ShimCometExpr` into one when we remove `spark-3.x` dir.

tshauck commented:
Hi, I updated the shim handling for 3.2 and made various other updates (based on PR feedback and general cleanup). Please have another look and let me know what you think, thanks!

viirya commented on spark/src/main/spark-3.2/org/apache/comet/shims/ShimCometUnhexExpr.scala:
```diff
@@ -0,0 +1,30 @@
+/**
+ * `ShimCometExpr` parses the `Unhex` expression assuming that the catalyst version is 3.2.x.
+ */
+trait ShimCometExpr {
```
The file name should be `ShimCometExpr.scala` instead of `ShimCometUnhexExpr.scala`?

andygrove replied:
Maybe the name could be `CometExprShim` instead?

viirya replied:
+1 for `CometExprShim`.

tshauck replied:
Thanks for the feedback, updated to `CometExprShim` in https://github.com/apache/datafusion-comet/pull/342/commits/a378f7433836d81543c4e6b8ce6c108934a893c5 along with updating the docstrings.

viirya commented on spark/src/main/spark-3.2/org/apache/comet/shims/ShimCometUnhexExpr.scala:
```diff
@@ -0,0 +1,30 @@
+import org.apache.spark.sql.catalyst.expressions._
+
+/**
+ * `ShimCometExpr` parses the `Unhex` expression assuming that the catalyst version is 3.2.x.
```
I think this trait will put all expr shims here. So maybe you can move this comment to `unhexSerde`.

tshauck replied:
Thanks, updated in https://github.com/apache/datafusion-comet/pull/342/commits/a378f7433836d81543c4e6b8ce6c108934a893c5.

viirya commented on pom.xml:
```diff
@@ -88,7 +88,8 @@ under the License.
     <additional.3_3.test.source>spark-3.3-plus</additional.3_3.test.source>
     <additional.3_4.test.source>spark-3.4</additional.3_4.test.source>
-    <shims.source>spark-3.x</shims.source>
+    <shims.majorSource>spark-3.x</shims.majorSource>
+    <shims.minorSource>spark-3.4</shims.minorSource>
```
Hmm, I don't get the naming? What `majorSource` and `minorSource` mean?

viirya replied:
How about `majorVerSrc` and `minorVerSrc`?

tshauck replied:
Fine by me, done in [112c7c6](https://github.com/apache/datafusion-comet/pull/342/commits/112c7c61db86f9a7383b0667af846ac02b1703c7)

viirya commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -1025,6 +1025,19 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
   }
 
+  test("unhex") {
+    Seq(false, true).foreach { dictionary =>
+      withSQLConf("parquet.enable.dictionary" -> dictionary.toString) {
```
Hmm, as there is only one value, does the `dictionary` actually work?

kazuyukitanimura commented on core/src/execution/datafusion/expressions/scalar_funcs/unhex.rs:
```diff
@@ -0,0 +1,149 @@
+
+    // Adjust the string if it has an odd length, and prepare to add a padding byte if needed.
+    let needs_padding = string.len() % 2 != 0;
+    let adjusted_string = if needs_padding { &string[1..] } else { string };
```
If I understand this correctly, `string[0]` is discarded when the length is odd, is it intentional?

andygrove replied:
Here is the logic in Spark 3.4.2 for handling the first char if the input is padded, for reference. It looks like there is some validation of the first digit that we do not have in this PR and it also looks like the unhexed digit is stored in the output is used in the return value if the length of the input string is 1. It would be good to make sure that we have tests covering this case.

```scala
    if ((bytes.length & 0x01) != 0) {
      // padding with '0'
      if (bytes(0) < 0) {
        return null
      }
      val v = Hex.unhexDigits(bytes(0))
      if (v == -1) {
        return null
      }
      out(0) = v
      i += 1
      oddShift = 1
    }
```

andygrove commented on common/pom.xml:
```diff
@@ -179,7 +179,8 @@ under the License.
               <sources>
-                <source>src/main/${shims.source}</source>
+                <source>src/main/${shims.majorSource}</source>
+                <source>src/main/${shims.minorSource}</source>
```
Thanks for adding the minor version shims.  These are going to help me with some of work around supporting `cast`.

andygrove reviewed (commented):
Thanks for the contribution @tshauck. This is looking great. I added some comments around additional testing.

tshauck commented:
Thanks for all the feedback. I think I've addressed the build/naming/etc feedback, and will have a look at improving the tests and any associated implementation changes sometime tomorrow. I'll request another review via GH when it's ready.

tshauck commented:
I think this is ready for review. I updated the `unhex` impl to be more faithful to Spark's (for odd-length inputs in particular), added better null handling, and added more tests from the Spark repo.

tshauck commented:
Err... looks to be an issue w/ spark 3.2 I'll need to look into. Hopefully the majority of the code'll remain unchanged.

<img width="1280" alt="image" src="https://github.com/apache/datafusion-comet/assets/421839/0ea2d7a2-0122-48a9-bb87-bb384d22e7ef">

andygrove commented:
Thanks for the updates @tshauck. I plan on reviewing later today.

andygrove commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -1024,7 +1024,25 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
+        |('A1B'),
+        |('0A1B')""".stripMargin)
 
+      checkSparkAnswerAndOperator(s"SELECT unhex(col) FROM $table")
```
It may be worth adding an `ORDER BY` clause here to ensure that the test is deterministic.

```suggestion
      checkSparkAnswerAndOperator(s"SELECT unhex(col) FROM $table ORDER BY col")
```

andygrove replied:
I wonder if the Spark 3.2 failure is at all related to ordering?


```
   [WrappedArray(10, 27)]                                  [WrappedArray(10, 27)]
  ![WrappedArray(115, 116, 114, 105, 110, 103)]            [WrappedArray(10, 27)]
  ![WrappedArray(27, 0)]                                   [WrappedArray(115, 116, 114, 105, 110, 103)]
  ```

tshauck replied:
Thanks, yeah, I looked at that bit. I think the test results are sorted, but the Spark case (left column) 27 sorts after 115.

Unfortunately, I think it may be legitimately different behavior between the 3.2 and the more recent versions.
<img width="995" alt="image" src="https://github.com/apache/datafusion-comet/assets/421839/36c59275-83b9-4099-9692-9913313f85c7"> .

tshauck replied:
Looking at the spark repo a bit more, it looks like there was a bug in the [3.2 implementation](https://github.com/apache/spark/commit/276abe3c3311300cffc6570b66f3977ea8172ff0) that was fixed in subsequent versions.

How should I handle this? Like should I write a native implementation of the erroneous code in order to be faithful to the spark, or have it be correct if different?

andygrove replied:
Actually adding the `ORDER BY` causes the test to fail because it no longer runs fully native, so ignore that suggestion.

Also, I do see code differences in Spark between 3.2 and 3.4 in the unhex algorithm. 3.2 does not have the `oddShift` variable. `oddShift` was added in https://github.com/apache/spark/commit/276abe3c3311300cffc6570b66f3977ea8172ff0

I guess the options are:

- mark `unhex` as incompat just for 3.2 and skip the test for 3.2 (probably the easiest path)
- implement per-spark-version logic in Rust for unhex

Let me know what you think or if you have any questions on this. This is a good example of the challenge of supporting multiple Spark versions 😓

tshauck replied:
Thanks for looking, I think we found the same commit w.r.t. 3.2's differences. 

I'll just go the easy route since matching 3.2 would be implementing buggy code and 3.2 is the oldest supported version (so presumably it'll be dropped at some point). I'll use this as a chance to poke around the repo, but may ask in discord if I get blocked on marking it incompat for 3.2 and skipping the tests for that version.

tshauck replied:
Made an update in https://github.com/apache/datafusion-comet/pull/342/commits/36baf8e87dd2689baf8c45961c7ed34a97f1f493 that seems to work (https://github.com/tshauck/arrow-datafusion-comet/actions/runs/8994978806). Happy to get any feedback about the approach taken.

andygrove commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -1024,7 +1024,27 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     }
   }
+  test("unhex") {
+    assume(!isSpark32, "unhex function has incorrect behavior in 3.2")
```
```suggestion
    // When running against Spark 3.2, we include a bug fix for https://issues.apache.org/jira/browse/SPARK-40924 that
    // was added in Spark 3.3, so although Comet's behavior is more correct when running against Spark 3.2, it is not
    // the same (and this only applies to edge cases with hex inputs with lengths that are not divisible by 2)
    assume(!isSpark32, "unhex function has incorrect behavior in 3.2")
```

andygrove replied:
follow up issue: https://github.com/apache/datafusion-comet/issues/400

andygrove commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -1024,7 +1024,27 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
+      sql(s"""INSERT INTO $table VALUES
+        |('537061726B2053514C'),
+        |('737472696E67'),
+        |('\\0'),
```
Removing this test fixes the Scala 2.13 build errors. I confirmed locally with `mvn clean package -DskipTests -Pscala-2.13`.

```suggestion
```

andygrove commented:
I plan on merging this tomorrow if there is no more feedback

kazuyukitanimura commented on spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:
```diff
@@ -1396,6 +1397,16 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde {
           val optExpr = scalarExprToProto("atan2", leftExpr, rightExpr)
           optExprWithInfo(optExpr, expr, left, right)
 
+        case e: Unhex if !isSpark32 =>
```
Unrelated, but potentially we can vote in the community when to deprecate 3.2 support...

kazuyukitanimura reviewed (commented):
Sorry, one more comment

tshauck commented on core/src/execution/datafusion/expressions/scalar_funcs/unhex.rs:
```diff
@@ -0,0 +1,229 @@
+                    } else if fail_on_error {
+                        return exec_err!("Input to unhex is not a valid hex string: {s}");
+                    } else {
+                        builder.append_null();
```
All good, thank you very much for looking. I think you're right in that there was a bug, which should be fixed and tested in https://github.com/apache/datafusion-comet/pull/342/commits/c5c3fcd51edddba44d1dd3b5ad176066ece0a5eb.

kazuyukitanimura reviewed (approved):
LGTM

kazuyukitanimura commented:
I forgot to ask about dictionary and scalar. Filed https://github.com/apache/datafusion-comet/issues/477
